### ISSUE TITLE
[Bug] Crash on Windows when switching scale_algorithm on 512-wide mode [sf#524]

### ISSUE THREAD

**Issue by andete** [openMSX/openMSX#836]
**Reported by manuelbi on 2014-03-09 21:42 UTC**
On Windows (both 32 and 64 bit) a crash happens in this scenario:
1) set renderer on SDL
2) set scale_factor on 2 or 3
3) start up an MSX and make sure you're on a 512-pixel wide mode (WIDTH 80 in text mode, or screen 6 or 7)
4) open OSD menu
5) go to video settings and keep the right or left cursor pressed on the scale algorithm setting to quickly loop through them.

**Comment by andete** [comment]
**Commented by manuelbi on 2014-03-11 12:21 UTC**
- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -1,6 +1,6 @@
 On Windows (both 32 and 64 bit) a crash happens in this scenario:
 1) set renderer on SDL
-2) set scale_factor on 3
+2) set scale_factor on 2 or 3
 3) start up an MSX and make sure you're on a 512-pixel wide mode (WIDTH 80 in text mode, or screen 6 or 7)
 4) open OSD menu
 5) go to video settings and keep the right or left cursor pressed on the scale algorithm setting to quickly loop through them. 

~~~~

**Comment by andete** [comment]
**Commented by manuelbi on 2014-03-11 16:45 UTC**
Stack trace from Vampier:

 	openmsx.exe!__chkstk() Line 109	Unknown
>	openmsx.exe!openmsx::MLAAScaler<unsigned int>::scaleImage(openmsx::FrameSource & src={...}, const openmsx::RawFrame * superImpose=0x0000000000000000, unsigned int srcStartY=24, unsigned int srcEndY=217, unsigned int srcWidth=640, openmsx::ScalerOutput<unsigned int> & dst={...}, unsigned int dstStartY=72, unsigned int dstEndY=651) Line 193	C++
 	openmsx.exe!openmsx::FBPostProcessor<unsigned int>::paint(openmsx::OutputSurface & output={...}) Line 326	C++
 	openmsx.exe!openmsx::Display::repaint(openmsx::OutputSurface & surface={...}) Line 402	C++
 	openmsx.exe!openmsx::Display::repaint() Line 385	C++
 	openmsx.exe!openmsx::Display::signalEvent(const std::shared_ptr<openmsx::Event const > & event={...}) Line 233	C++
 	openmsx.exe!openmsx::EventDistributor::deliverEvents() Line 109	C++
 	openmsx.exe!openmsx::Reactor::run(openmsx::CommandLineParser & parser={...}) Line 612	C++
 	openmsx.exe!openmsx::SDL_main(int argc=1, char * * argv=0x000000e53899fa10) Line 153	C++
 	openmsx.exe!SDL_main(int argc=1, char * * argv=0x000000e53896bbc0) Line 182	C++
 	openmsx.exe!main(int argc=1, char * * argv=0x000000e53896bbc0) Line 315	C
 	openmsx.exe!__tmainCRTStartup() Line 255	C
 	openmsx.exe!mainCRTStartup() Line 165	C
 	kernel32.dll!00007ff8332415cd()	Unknown
 	ntdll.dll!00007ff8355943d1()	Unknown



and it's this line: VLA(unsigned, verticals, srcNumLines * srcWidth);

So it seems restricted to MLAA (?).

**Comment by andete** [comment]
**Commented by manuelbi on 2014-03-11 19:46 UTC**
Fixed with commit 3774f000ca1d621acf32b1a2412f4a59f044fc52
Author: m9710797 <vermaelen.wouter@gmail.com>
Date:   Tue Mar 11 18:54:18 2014 +0100

    Fix(?) #524 Crash on Windows when switching scale_algorithm
    
    Vampier created a backtrace of this crash (thanks!). It points to this line:
      VLA(unsigned, verticals, srcNumLines * srcWidth);
    The strange thing is that a bit earlier in the code there is an almost
    identical VLA() for 'horizontals' and that works fine. Though I *think* the
    problem is the following:
    
    The MLAAScaler::scaleImage() method allocates quite a few large buffers on t
    stack. The size of these buffers is proportional to the amount of to-be-scal
    MSX pixels. In case of an high-resolution MSX image (screen0/80, screen6,
    screen7) The total required buffer space is more than 1MB! And the line abov
    is the allocation that pushes the stack size over 1MB.
    
    On windows the default stack size is only 1MB. On Linux it's typically 8MB. 
    this may explain why it only crashes on windows and only on hi-res MSX scree
    modes.

**Comment by andete** [comment]
**Updated by manuelbi on 2014-03-11 19:46 UTC**
- **status**: open --> closed-fixed
- **assigned_to**: Wouter Vermaelen
